src/Nethermind/Nethermind.Optimism/CL/Decoding/FrameDecoder.cs

```diff
@@ -2,64 +2,110 @@
-        {
-            ChannelId = channelId,
-            FrameNumber = frameNumber,
-            FrameData = frameData,
-            IsLast = isLast == 1
-        }, 23 + (int)frameDataLength);
     }
+}
+
+/// <remarks>
+/// https://specs.optimism.io/protocol/derivation.html#frame-format
+/// </remarks>
+public struct Frame : IEquatable<Frame>
+{
+    public UInt128 ChannelId;
+    public UInt16 FrameNumber;
+    public byte[] FrameData;
+    public bool IsLast;
 
-    public static Frame[] DecodeFrames(byte[] data)
+    public int Size
     {
-        byte version = data[0];
-        if (version != 0)
+        get
         {
-            throw new Exception($"Frame Decoder version {version} is not supported.");
+            unsafe
+            {
+                return sizeof(UInt128) + sizeof(UInt16) + sizeof(UInt32) + FrameData.Length + sizeof(byte);
+            }
         }
+    }
+
+    public bool Equals(Frame other)
+    {
+        return ChannelId.Equals(other.ChannelId) && FrameNumber == other.FrameNumber && FrameData.SequenceEqual(other.FrameData) && IsLast == other.IsLast;
+    }
 
-        List<Frame> frames = new List<Frame>();
-        int pos = 1;
-        while (pos < data.Length)
+    public override bool Equals(object? obj)
+    {
+        return obj is Frame other && Equals(other);
+    }
+
+    public override int GetHashCode()
+    {
+        return HashCode.Combine(ChannelId, FrameNumber, FrameData, IsLast);
+    }
+
+    public static int FromBytes(ReadOnlySpan<byte> buffer, out Frame frame)
+    {
+        unsafe
         {
-            (Frame frame, int decoded) = DecodeFrame(data[pos..]);
-            pos += decoded;
-            frames.Add(frame);
+            UInt128 channelId = BinaryPrimitives.ReadUInt128BigEndian(buffer.TakeAndMove(sizeof(UInt128)));
+            UInt16 frameNumber = BinaryPrimitives.ReadUInt16BigEndian(buffer.TakeAndMove(sizeof(UInt16)));
+            UInt32 frameDataLength = BinaryPrimitives.ReadUInt32BigEndian(buffer.TakeAndMove(sizeof(UInt32)));
+            ReadOnlySpan<byte> frameData = buffer.TakeAndMove((int)frameDataLength);
+            byte isLast = buffer.TakeAndMove(1)[0];
+            if (isLast != 0 && isLast != 1)
+            {
+                throw new FormatException($"Invalid {nameof(IsLast)} flag");
+            }
+
+            frame = new Frame
+            {
+                ChannelId = channelId,
+                FrameNumber = frameNumber,
+                // TODO: Can we avoid this copying?
+                FrameData = frameData.ToArray(),
+                IsLast = isLast == 1
+            };
+
+            return frame.Size;
         }
+    }
+
+    public int WriteTo(Span<byte> span)
+    {
+        int initialLength = span.Length;
 
-        if (pos != data.Length)
+        unsafe
         {
-            throw new Exception("Excess frame data");
+            BinaryPrimitives.WriteUInt128BigEndian(span.TakeAndMove(sizeof(UInt128)), ChannelId);
```

Comment:
I'd compare whether this or `span.Slice` with a fixed offset is faster.

Reply:
You mean the `span.TakeAndMove`?

Reply:
No, I meant `span.Slice` as we know the offsets here. No need, beside some consumption design, to move anything if you can slice off to the given fixed offset.